### PR TITLE
[FLINK-14400] Shrink scope of MemoryManager from TaskExecutor to slot

### PR DESCRIPTION
**Pull request by azagrebin**
## What is the purpose of the change

`MemoryManager` currently manages the memory bookkeeping for all slots/tasks inside one `TaskExecutor`. For better abstraction and isolation of slots, we can shrink its scope and make it per slot. The memory limits are fixed now per slot at the moment of slot creation. All operators, sharing the slot, will get their fixed fractional limits.

In future, we might make it possible for operators to over-allocate beyond their fraction limit if there is some available free memory in the slot but it should be possible to reclaim the over-allocated memory at any time if other operator decides to claim its fair share within its limit.

## Brief change log

  - introduce `TaskSlotTableBuilder` for tests
  - Refactor out slots creation from the `TaskSlotTable` constructor to the factory method in `TaskManagerServices`
  - `TaskSlot` has its own memory manager and `close` method to shutdown the memory manager and release all left unreleased memory (with a warning check whether everything has been already released)
  - When the `TaskSlotTable` is stopped, all slots are closed as well. Currently, it basically happens only if the TM is completely stoped, same as for the former TM's global memory manager
  - Remove the global memory manager from `TaskManagerServices`, keep only memory size
  - Remove now redundant number of slots from the `MemoryManager`.
  - Adjust unit tests

## Verifying this change

Existing unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by azagrebin** [comment]
currently based on #9910

File: flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java [hunk 1]
```diff
@@ -604,7 +582,7 @@ public int computeNumberOfPages(double fraction) {
 		}
 
 		//noinspection NumericCastThatLosesPrecision
-		return (int) (budgetByType.maxTotalNumberOfPages() * fraction / numberOfSlots);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java]: looooks like it is in fact used? 😕

File: flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java [hunk 2]
```diff
@@ -192,7 +171,6 @@ public boolean isShutdown() {
 	 *
 	 * @return True, if the memory manager is empty and valid, false if it is not empty or corrupted.
 	 */
-	@VisibleForTesting
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java]: doesn't belong in this commit

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServicesConfiguration.java [hunk 3]
```diff
@@ -115,6 +116,8 @@ public TaskManagerServicesConfiguration(
 		this.maxJvmHeapMemory = maxJvmHeapMemory;
 		this.localRecoveryEnabled = checkNotNull(localRecoveryEnabled);
 		this.queryableStateConfig = queryableStateConfig;
+
+		Preconditions.checkArgument(0 < numberOfSlots, "The number of task slots must be greater than 0.");
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServicesConfiguration.java]: The commit message says the check was `moved`, but this is just one more check.
Also the null check below doesn't make sense.

File: flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java [hunk 4]
```diff
@@ -89,6 +89,18 @@
 	/** Flag whether the close() has already been invoked. */
 	private volatile boolean isShutDown;
 
+	/**
+	 * Creates a memory manager with the given memory types, capacity and given page size.
+	 *
+	 * @param memorySizeByType The total size of the memory to be managed by this memory manager for each type (heap / off-heap).
+	 * @param pageSize The size of the pages handed out by the memory manager.
+	 */
+	public MemoryManager(
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java]: is github messing up the commit order again? to me it shows that the numberOfSlots argument is removed in a prior commit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java [hunk 5]
```diff
@@ -404,8 +404,6 @@ private void stopTaskExecutorServices() throws Exception {
 			exception = ExceptionUtils.firstOrSuppressed(e, exception);
 		}
 
-		taskSlotTable.stop();
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: where is this now called?

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableBuilder.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/** Builder for {@link TaskSlotTable}. */
+public class TaskSlotTableBuilder {
+	private static final long DEFAULT_SLOT_TIMEOUT = 10000L;
+
+	private static final ResourceProfile DEFAULT_RESOURCE_PROFILE =
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableBuilder.java]: why are we now using a different default profile? Previously, it was ResourceProfile.ANY, which also seems to be a safer default.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableBuilder.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+		return this;
+	}
+
+	public TaskSlotTableBuilder setTimerService(TimerService<AllocationID> timerService) {
+		this.timerService = timerService;
+		return this;
+	}
+
+	public TaskSlotTableBuilder withTimerServiceTimeout(Time timeout) {
+		this.timerService = new TimerService<>(TestingUtils.defaultExecutor(), timeout.toMilliseconds());
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableBuilder.java]: I'd rather have a dedicated factory method for creating such a timer service outside of this class which is then passed to the setter.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableBuilder.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+	private static TimerService<AllocationID> createDefaultTimerService() {
+		return new TimerService<>(TestingUtils.defaultExecutor(), DEFAULT_SLOT_TIMEOUT);
+	}
+
+	public static TaskSlotTableBuilder newBuilder() {
+		return newBuilderWithDefaultSlots(1);
+	}
+
+	public static TaskSlotTableBuilder newBuilderWithDefaultSlots(int numberOfDefaultSlots) {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableBuilder.java]: not sure about the value of this method; it's even longer then setting the slots explicitly.

`TaskManagerServices.newBuilder().setTaskSlots(2)`
`TaskManagerServices.newBuilderWithDefaultSlots(2)`

It would make sense to be if it returned a TaskSlotTable directly; as a short-hand to create a TST with X slots, which is the primary usage anyway.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableBuilder.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+		return new TaskSlotTableBuilder().setTaskSlots(createDefaultSlotProfiles(numberOfDefaultSlots));
+	}
+
+	public static List<ResourceProfile> createDefaultSlotProfiles(int numberOfDefaultSlots) {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableBuilder.java]: The naming of this method and argument doesn't really make sense if it is used outside of this class (like in the TaskSlotTableTest). I'd move it somewhere else and remove the `Default` part.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java [hunk 10]
```diff
@@ -82,28 +81,15 @@
 	private boolean started;
 
 	public TaskSlotTable(
-		final Collection<ResourceProfile> resourceProfiles,
+		final List<TaskSlot> taskSlots,
 		final TimerService<AllocationID> timerService) {
 
-		int numberSlots = resourceProfiles.size();
-
-		Preconditions.checkArgument(0 < numberSlots, "The number of task slots must be greater than 0.");
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java]: does this precondition no longer apply?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java [hunk 11]
```diff
@@ -302,6 +299,30 @@ public static TaskManagerServices fromConfiguration(
...
+			long timerServiceShutdownTimeout) {
+		final List<ResourceProfile> resourceProfiles =
+			Collections.nCopies(numberOfSlots, computeSlotResourceProfile(numberOfSlots, managedMemorySize));
+		final TimerService<AllocationID> timerService = new TimerService<>(
+			new ScheduledThreadPoolExecutor(1),
+			timerServiceShutdownTimeout);
+		return new TaskSlotTable(createTaskSlotsFromResources(resourceProfiles), timerService);
+	}
+
+	public static List<TaskSlot> createTaskSlotsFromResources(Collection<ResourceProfile> resourceProfiles) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: why is this public?

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/aggregate/BytesHashMapTest.java [hunk 12]
```diff
@@ -98,7 +98,6 @@ public void testHashSetMode() throws IOException {
 		MemoryManager memoryManager = MemoryManagerBuilder
 			.newBuilder()
 			.setMemorySize(numMemSegments * PAGE_SIZE)
-			.setNumberOfSlots(32)
```
Comment (zentol) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/aggregate/BytesHashMapTest.java]: I would merge this commit with the one removing the numberOfSlots argument. Makes it much clearer that this commit does affect behavior.

File: flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java (hunk 1 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java]: as it was supposed to be the last commit :), at this point `numberOfSlots` can be only 1, so it is needed anymore

File: flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java (hunk 4 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java]: I will push again, hopefully it will fix the commit order, otherwise I will add commit order into the PR description

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 5 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: It's always been also part of `taskExecutorServices.shutDown();`, further below

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableBuilder.java (hunk 6 above)
Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableBuilder.java]: I want to avoid creation of slot `MemoryManager` with infinite memory. It is just a dummy memory size anyways by default.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableBuilder.java (hunk 9 above)
Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableBuilder.java]: let's refactor `TaskSlotTableBuilder` to `TaskSlotAndTableFactory`.
That should also resolve other comments for this commit

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java (hunk 10 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java]: added back

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java (hunk 11 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: used in `TaskSlotTableBuilder`, true, not needed

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java (hunk 10 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java]: added back

**Comment by azagrebin** [comment]
Thanks for the review @zentol, addressed comments
The commit order looks correct now.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotAndTableFactory.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+import org.apache.flink.runtime.clusterframework.types.AllocationID;
+import org.apache.flink.runtime.clusterframework.types.ResourceProfile;
+import org.apache.flink.runtime.memory.MemoryManager;
+import org.apache.flink.runtime.testingUtils.TestingUtils;
+
+import java.util.Collections;
+import java.util.List;
+
+/** Factory for {@link TaskSlotTable} and {@link TaskSlot}s. */
+public enum TaskSlotAndTableFactory {
```
[the remark this answers is not in the document]
Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotAndTableFactory.java]: `ResourceTestUtils` is probably too broad but `TaskSlotUtils` would make sense imo

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java [hunk 14]
```diff
@@ -258,15 +253,15 @@ public static TaskManagerServices fromConfiguration(
 			taskManagerServicesConfiguration.getTaskManagerAddress(),
 			dataPort);
 
-		// this call has to happen strictly after the network stack has been initialized
-		final MemoryManager memoryManager = createMemoryManager(taskManagerServicesConfiguration);
-
 		final BroadcastVariableManager broadcastVariableManager = new BroadcastVariableManager();
 
+		// this call has to happen strictly after the network stack has been initialized
```
[the remark this answers is not in the document]
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: This looks like a stale comment.
Originally, it looks like to be introduced for network pre-allocating direct memory first (another comment, already deleted). Now memory manager does not pre-allocate anything anymore, so they should interfere with each other. Anyways the memory usage should all fit together at the end without overlapping otherwise it is a bug.
I added a hotfix commit to remove the confusing comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java [hunk 15]
```diff
@@ -305,4 +321,23 @@ public String toString() {
 		return "TaskSlot(index:" + index + ", state:" + state + ", resource profile: " + resourceProfile +
 			", allocationId: " + (allocationId != null ? allocationId.toString() : "none") + ", jobId: " + (jobId != null ? jobId.toString() : "none") + ')';
 	}
+
+	@Override
+	public void close() {
+		verifyMemoryFreed();
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java]: are there valid use-cases where this assumption is violated? If not, how about adding an assert in `verifyMemoryFreed`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java [hunk 16]
```diff
@@ -1459,7 +1420,7 @@ public void testReconnectionAttemptIfExplicitlyDisconnected() throws Exception {
 	 */
 	@Test
 	public void testInitialSlotReport() throws Exception {
-		final TaskSlotTable taskSlotTable = new TaskSlotTable(Collections.singleton(ResourceProfile.ANY), timerService);
+		final TaskSlotTable taskSlotTable = TaskSlotUtils.createTaskSlotTable(1);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: why are you no longer passing the timerService? (also applies to many other test methods)

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java [hunk 17]
```diff
@@ -490,14 +497,17 @@ private static void checkTempDirs(String[] tmpDirs) throws IOException {
...
+	private static ResourceProfile computeSlotResourceProfile(int numOfSlots, Map<MemoryType, Long> memorySizeByType) {
 		return new ResourceProfile(
 			Double.MAX_VALUE,
 			MemorySize.MAX_VALUE,
 			MemorySize.MAX_VALUE,
-			// TODO: before operators separate on-heap/off-heap managed memory, we use on-heap managed memory to denote total managed memory
-			MemorySize.parse(managedMemoryPerSlotMB + "m"),
-			MemorySize.MAX_VALUE,
+			new MemorySize(memorySizeByType.getOrDefault(MemoryType.HEAP, 0L) / numOfSlots),
+			new MemorySize(memorySizeByType.getOrDefault(MemoryType.OFF_HEAP, 0L) / numOfSlots),
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: Why is it not problematic that the RM always believes that a worker has no off-heap memory available? (The RM calls `computeSlotResourceProfile(int numOfSlots, long managedMemorySize)` which never sets the OFF_HEAP memory type, hence it being 0)

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java [hunk 18]
```diff
@@ -490,14 +497,17 @@ private static void checkTempDirs(String[] tmpDirs) throws IOException {
 	}
 
 	public static ResourceProfile computeSlotResourceProfile(int numOfSlots, long managedMemorySize) {
-		int managedMemoryPerSlotMB = (int) bytesToMegabytes(managedMemorySize / numOfSlots);
+		// TODO: before operators separate on-heap/off-heap managed memory, we use on-heap managed memory to denote total managed memory
+		return computeSlotResourceProfile(numOfSlots, Collections.singletonMap(MemoryType.HEAP, managedMemorySize));
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: I'm wondering if we shouldn't in-line this call into the ResourceManager, with a dedicated comment as to why setting only the HEAP memory type isn't a problem. It's a bit weird to have a method in this class if it never even called in the context of a task executor.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java (hunk 17 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: At the moment, our configuration contains only one size of managed memory and it can be only of one configured type but internal spec/profile data structures are already prepared to have two sizes for both types of memory. When the main PR #10161 of FLIP-49 is merged, we will switch everything into having two types of memory.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java (hunk 18 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: True, we could but it should completely disappear soon after #10161.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java (hunk 16 above)
Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: they all use the default timer service, same as in the builder by default, which is closed when TM terminates and slot table is stopped. I will add a hotfix commit to remove this test class field `timerService` as it is redundant after this refactoring.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java (hunk 15 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java]: I cannot come up with any valid use-cases for this check to fail and firstly I had an assertion. I had some failing test cases previously (probably for some new blink code) but CI does not fail anymore, they might have already been fixed. I tried CI with a commit inserting the assertion: https://travis-ci.org/azagrebin/flink/builds/613505573. It is also green.

Anyways, as discussed offline, memory leak after finishing task should not fail the whole TM but we have to provide a warning.
